text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,544 @@
+
+## Detailed design
+
+Every registry API will be updated to take a new `Identifier` type in place of the string microsyntax, while maintaining backwards compatibility by introducing versioning to the resolver. A codemod will be provided to allow users to migrate directly to the new API. The legacy resolver APIs are deprecated until Ember 4.0.
+
+(TypeScript consumers should see [**Proposed Type Definitions**](#proposed-type-definitions) in the appendix for additional details on type safety and resolution.)
+
+### `Identifier`
+
+The core new type in the design of the API is an *identifier*, which is always passed as the first argument to resolver APIs.
+
+```ts
+interface Identifier {
+  type: string;
+  name: string;
+  namespace?: string;
+}
+```
+
+Here the `type` corresponds to the prefix component of the legacy API, and `name` to the postfix component of the legacy API: `service:foo` becomes `{ type: 'service', name: 'foo' }`.
+
+The `namespace` field allows for identifiers to be restricted to a particular namespace, as in for example tools which namespace lookups for addons (e.g. [ember-holy-futuristic-template-namespacing-batman](https://github.com/rwjblue/ember-holy-futuristic-template-namespacing-batman)). It is optional since it is not required in normal usage.
+
+We also introduce `FactoryTypeIdentifier` to distinguish between injections for *all factories of a given type* (`FactoryTypeIdentifier`) and *specific factories* (`Identifier`):
+
+```ts
+interface FactoryTypeIdentifier {
+  type: string;
+}
+```
+
+As is the case in the microsyntax-based design, these factory type injections may not be namespaced.
+
+### `Resolver`
+
+Ember’s `Resolver` function is public API, designed to be customized and overridden. To support backwards compatibility with existing custom resolvers, we introduce a `schemaVersion` key to the `Resolver` API, which may be `undefined` or an integer value. If `schemaVersion` is `undefined` or `0`, the legacy behavior of the resolver is maintained:
+
+```ts
+interface ResolverV0 {
+  schemaVersion?: 0;
+  resolve(fullName: string);
+}
+```
+
+If `schemaVersion` is `1`, the `resolve` function has a new signature, using the `Identifier` type introduced above:
+
+```ts
+interface ResolverV1 {
+  schemaVersion: 1;
+  resolve(identifier: Identifier)
+}
+```
+
+The public API is *either* of the Resolver variants:
+
+```ts
+type Resolver = ResolverV0 | ResolverV1;
+```
+
+The other registry API changes may also be implemented in terms of the new resolver `schemaVersion` check.
+
+This design is completely backwards compatible: it continues working exactly as it does today, with users opting *into* the change by upgrading their version of `ember-resolver`, which will implement the new design proposed by this RFC as a breaking change.
```

Comment:
I think this is conceptually fine (making the upgrade of ember-resolver a breaking change update), but I think it would actually be fine to do compatibly.

Something like changing [this line](https://github.com/ember-cli/ember-resolver/blob/8f624f294eb4a2c77bb40e619d7cf347f63ecd17/addon/resolvers/classic/index.js#L150) to:

```js
resolve(fullNameOrIdentifier) {
  let parsedName = typeof fullNameOrIdentifier === 'string' ? this.parseStringName(fullNameOrIdentifier) : this.parseIdentifier(fullNameOrIdentifier);
  // ...snip...
}
```

Reply:
Interesting! I think I *prefer* the breaking change just for the clarity, but perhaps we'd be better to do the same thing here as in general: introduce in a compatible way, but deprecate, and then drop in a later major?

Reply:
Incorporated the suggestions here in e0efbc56!